atlasdb-config/src/main/java/com/palantir/atlasdb/factory/AtlasDbDialogueServiceProvider.java

```diff
@@ -0,0 +1,142 @@
+import com.palantir.conjure.java.api.config.service.UserAgent;
+import com.palantir.conjure.java.client.config.ClientConfiguration;
+import com.palantir.conjure.java.client.config.NodeSelectionStrategy;
+import com.palantir.dialogue.Channel;
+import com.palantir.dialogue.ConjureRuntime;
+import com.palantir.dialogue.clients.DialogueClients;
+import com.palantir.lock.client.DialogueAdaptingConjureTimelockService;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.refreshable.Refreshable;
+
+/**
+ * Provides a mechanism for accessing services that use Dialogue for communication. A service is defined as a cluster of
+ * zero or more nodes, where contacting any of the nodes is legitimate (subject to redirects via 308s and 503s). If
+ * working with heterogeneous nodes and/or broadcast is important (e.g. for Paxos Acceptor use cases), you should be
+ * VERY careful when using this class.
+ *
+ * Proxies must be resilient to servers repeatedly returning 308s that are large in number, but persist for only a short
+ * duration. Furthermore, proxies should include in their {@link com.palantir.conjure.java.api.config.service.UserAgent}
+ * information to allow client services to identify the protocol they are using to talk, via
+ * {@link AtlasDbHttpProtocolVersion}.
+ */
+public final class AtlasDbDialogueServiceProvider {
+    private static final String TIMELOCK_SHORT_TIMEOUT = "timelock-short-timeout";
+    private static final String TIMELOCK_LONG_TIMEOUT = "timelock-long-timeout";
+    private final DialogueClients.ReloadingFactory dialogueClientFactory;
+
+    private AtlasDbDialogueServiceProvider(DialogueClients.ReloadingFactory dialogueClientFactory) {
+        this.dialogueClientFactory = dialogueClientFactory;
+    }
+
+    public static AtlasDbDialogueServiceProvider create(
+            Refreshable<ServerListConfig> timeLockServerListConfig,
+            DialogueClients.ReloadingFactory baseFactory,
+            UserAgent userAgent) {
+        UserAgent versionedAgent = userAgent.addAgent(AtlasDbRemotingConstants.ATLASDB_HTTP_CLIENT_AGENT);
+        Refreshable<Map<String, RemoteServiceConfiguration>> timeLockRemoteConfigurations = timeLockServerListConfig
+                .map(serverListConfig -> ImmutableMap.of(
+                        TIMELOCK_SHORT_TIMEOUT,
+                        createRemoteServiceConfiguration(versionedAgent, serverListConfig, false),
+                        TIMELOCK_LONG_TIMEOUT,
+                        createRemoteServiceConfiguration(versionedAgent, serverListConfig, true)));
+        DialogueClients.ReloadingFactory reloadingFactory
+                = decorateForFailoverServices(baseFactory, timeLockRemoteConfigurations).withUserAgent(versionedAgent);
+
+        return new AtlasDbDialogueServiceProvider(reloadingFactory);
+    }
+
+    ConjureTimelockService getConjureTimelockService() {
+        Preconditions.checkState(isDialogue(ConjureTimelockServiceBlocking.class),
```

Comment:
Let's kill this check because it's really hacky and currently not needed (and let's keep it that way)

Reply:
Yep, this was needed for the other design / in general designs that allow users to pass in `Class<T>` but if we're controlling it it should be fine.